pkg/controller/mighook/validation.go

```diff
@@ -0,0 +1,125 @@
+package mighook
+
+import (
+	"github.com/docker/distribution/reference"
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+// Types
+const (
+	InvalidTargetCluster = "InvalidTarget"
+	InvalidImage         = "InvalidImage"
+	InvalidNSName        = "InvalidNSName"
+	InvalidSAName        = "InvalidSAName"
+)
+
+// Categories
+const (
+	Critical = migapi.Critical
+)
+
+// Reasons
+const (
+	Supported    = "Supported"
+	NotSupported = "NotSupported"
+	NotSet       = "NotSet"
+	NotFound     = "NotFound"
+	KeyError     = "KeyError"
+	TestFailed   = "TestFailed"
+)
+
+// Statuses
+const (
+	True  = migapi.True
+	False = migapi.False
+)
+
+// Messages
+const (
+	ReadyMessage                = "The hook is ready."
+	InvalidTargetClusterMessage = "Only 'source' and 'destination' are accepted as values for spec.targetCluster."
+	InvalidImageMessage         = "The image name specified in spec.Image is an invalid."
+	InvalidNSNameMessage        = "The executionNamespace specified is invalid."
+	InvalidSANameMessage        = "The serviceAccount specified is invalid."
+)
+
+// Validate the hook resource.
+func (r ReconcileMigHook) validate(hook *migapi.MigHook) error {
+	err := r.validateNS(hook)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+	err = r.validateImage(hook)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+	err = r.validateSA(hook)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+	err = r.validateTargetCluster(hook)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	return nil
+}
+
+func (r ReconcileMigHook) validateNS(hook *migapi.MigHook) error {
+	if errs := validation.IsDNS1123Label(hook.Spec.ExecutionNamespace); len(errs) != 0 {
+		hook.Status.SetCondition(migapi.Condition{
+			Type:     InvalidNSName,
+			Status:   True,
+			Reason:   NotSet,
+			Category: Critical,
+			Message:  InvalidNSNameMessage,
+		})
+	}
+	return nil
+}
+
+func (r ReconcileMigHook) validateImage(hook *migapi.MigHook) error {
+	match := reference.ReferenceRegexp.MatchString(hook.Spec.Image)
```

Comment:
This appears to bring in docker `distribution` as a dep just to get access to an a valid image reference regex. I wonder if this is heavy handed, and how hard it would be to construct the regex within this function.

See https://github.com/docker/distribution/blob/master/reference/regexp.go#L65-L70

Reply:
Maybe, but that brings in at least four functions anchored, literal, capture and optional, several vars NameRexexp which in turn pulls in DomainRegexp and nameComponentRegexp, domainRegexp means we need domainComponentRegexp, nameComponentRegexp needs alphaNumericRegexp... we also have TagRegexp and DigestRegexp.

I'm not eager to recreate this crazy logic.

Reply:
Looking closer the regexp.go in the docker/distribution reference pkg does not require any of the additional deps being pulled in. The file pretty much stands on its own and the rest of the stuff pulled in is a consequence of other stuff in their reference pkg. Conveiently, we also have a reference pkg. I dropped it there and imported from there and it worked fine. Last PR removes docker distribution vendor stuff. Don't know if that's more acceptable.